### PR TITLE
Update the geofencing subscription models to align on CAMARA commonalities

### LINKED ISSUE #118: Geofencing API - Add "format: uri" to notificationUrl
**Problem description**
To have a strong typed notificationUrl, add a format constraint on type. 
This will simplify implementation and clarify usage.
```yaml
    NotificationUrl:
      type: string
      format: uri
      example: "https://application-server.com"
      description: https callback address where the event notification must be POST-ed
```

### LINKED ISSUE #124: Geofencing API - Defining a standard behavior for first event
**Problem description**
Hello
Looking to define a standard behavior within our implementations.
If we got a suscription for ``org.camaraproject.geofencing.v0.area-entered`` event type for a device **already** present in the geo zone our understanding is to **not** trigger an event.  
Same for ``org.camaraproject.geofencing.v0.area-left`` - if the device is not in the monitored zone we do not sent notification.

If for any reason the device is not localized at the time of the subscription we apply same rule

Given the event name seems that this is the expected behavior but looking for team feedback. 

**Possible evolution**
A short description of this rule in the documentation

**Alternative solution**

**Additional context**

### LINKED ISSUE #200: Geofencing - Align with new model
**Problem description**
We have decided some months ago to move to CloudsEvent model for subscription. Now, the documentation is available in commonalities, in the [design document](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-design-guidelines.md#121-subscriptionl), and with a [yaml template](https://github.com/camaraproject/Commonalities/blob/main/artifacts/camara-cloudevents/event-subscription-template.yaml).

Our Geofencing API should now be updated accordingly to this new model.

**Expected behavior**

Align geofencing API with this model.
This move will de facto solve #124 

**Alternative solution**

**Additional context**

@maxl2287 as you did a great job on this API could we expect you to continue the work for this issue?

### PR DESCRIPTION
**Pull request by maxl2287**
#### What type of PR is this?

* correction
* enhancement/feature

#### What this PR does / why we need it:
Update of the subscription-models based on: 
- https://github.com/camaraproject/Commonalities/blob/main/documentation/API-design-guidelines.md#12-subscription-notification--event
- https://github.com/camaraproject/Commonalities/blob/main/artifacts/camara-cloudevents/event-subscription-template.yaml

#### Fixed Issues
- Fixes #124 
- Fixes #118 
- Fixes #200

### CODE REVIEW

**Comment by maxl2287** [comment]
@bigludo7 @jlurien 

What are we doing about this here?
Are we setting here a maximum of 1 for now?
Because we are not yet supporting multiple types, as this would be maybe an own PR / issue.

Wdyt?

```yaml
        types:
          description: |
            Camara Event types eligible to be delivered by this subscription.
            Note: for the Commonalities meta-release v0.4 we enforce to have only event type per subscription then for following meta-release use of array MUST be decided
             at API project level.
          type: array
          items:
            $ref: "#/components/schemas/SubscriptionEventType"
```

**Comment by bigludo7** [comment]
> @bigludo7 @jlurien
> 
> What are we doing about this here? Are we setting here a maximum of 1 for now? Because we are not yet supporting multiple types, as this would be maybe an own PR / issue.
> 
> Wdyt?
> 
> ```yaml
>         types:
>           description: |
>             Camara Event types eligible to be delivered by this subscription.
>             Note: for the Commonalities meta-release v0.4 we enforce to have only event type per subscription then for following meta-release use of array MUST be decided
>              at API project level.
>           type: array
>           items:
>             $ref: "#/components/schemas/SubscriptionEventType"
> ```

I think we can add this `maximum:1` in our yaml. The template in commonalities is generic and should live for more than one release. Here we're crafting an yaml for implementation so better to be straightforward.
@jlurien @alpaycetin74 different view?

**Review by bigludo7 — DISMISSED** [review]
LGFM !
Thanks Max

**Comment by jlurien** [comment]
> > @bigludo7 @jlurien
> > What are we doing about this here? Are we setting here a maximum of 1 for now? Because we are not yet supporting multiple types, as this would be maybe an own PR / issue.
> > Wdyt?
> > ```yaml
> >         types:
> >           description: |
> >             Camara Event types eligible to be delivered by this subscription.
> >             Note: for the Commonalities meta-release v0.4 we enforce to have only event type per subscription then for following meta-release use of array MUST be decided
> >              at API project level.
> >           type: array
> >           items:
> >             $ref: "#/components/schemas/SubscriptionEventType"
> > ```
> 
> I think we can add this `maximum:1` in our yaml. The template in commonalities is generic and should live for more than one release. Here we're crafting an yaml for implementation so better to be straightforward. @jlurien @alpaycetin74 different view?

I agree to setting the maximum to 1 in this version as that will impose implementations to follow the agreed current limit.

**Review by jlurien — COMMENTED** [review]
I think we should include in our spec for this version the values that are allowed: protocol=HTTP, credentialType=ACCESSTOKEN and remove the content for all other options which are not applicable.

File: code/API_definitions/geofencing-subscriptions.yaml [hunk 1]
```diff
@@ -278,16 +300,23 @@ paths:
...
-
 components:
   securitySchemes:
     openId:
       type: openIdConnect
       openIdConnectUrl: https://example.com/.well-known/openid-configuration
+    notificationsBearerAuth:
+      type: http
+      scheme: bearer
+      bearerFormat: "{$request.body#/sinkCredential.credentialType}"
```
[the remark this answers is not in the document]
Comment (bigludo7) [reply code/API_definitions/geofencing-subscriptions.yaml]: @jlurien Just checked...Not coming from the cloudEvents [spec](https://github.com/cloudevents/spec/blob/main/subscriptions/subscriptions-openapi.yaml) but we have it in our [template](https://github.com/camaraproject/Commonalities/blob/main/artifacts/camara-cloudevents/event-subscription-template.yaml).
Need to check with... https://github.com/PedroDiez why we have it in the template ;)

File: code/API_definitions/geofencing-subscriptions.yaml [hunk 2]
```diff
@@ -47,21 +47,23 @@ info:
     It is used when:
       - the subscription expire time (optionally set by the requester) has been reached
       - the maximum number of subscription events (optionally set by the requester) has been reached
+      - the Access Token sinkCredential (optionally set by the requester) expiration time has been reached
```
[the remark this answers is not in the document]
Comment (bigludo7) [reply code/API_definitions/geofencing-subscriptions.yaml]: @jlurien This is mentioned in the guideline in this [§](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-design-guidelines.md#resource-based-explicit-subscription) after the sinkCredential table.

I like your rule proposal  between config.subscriptionExpireTime and sinkCredential.accessTokenExpireUtc. Probably a good 'tips' for commonalities guideline.

**Comment by jlurien** [comment]
FYI, @PedroDiez, as you have been more involved than me in this track in Commonalities

**Comment by alpaycetin74** [comment]
> > @bigludo7 @jlurien
> > What are we doing about this here? Are we setting here a maximum of 1 for now? Because we are not yet supporting multiple types, as this would be maybe an own PR / issue.
> > Wdyt?
> > ```yaml
> >         types:
> >           description: |
> >             Camara Event types eligible to be delivered by this subscription.
> >             Note: for the Commonalities meta-release v0.4 we enforce to have only event type per subscription then for following meta-release use of array MUST be decided
> >              at API project level.
> >           type: array
> >           items:
> >             $ref: "#/components/schemas/SubscriptionEventType"
> > ```
> 
> I think we can add this `maximum:1` in our yaml. The template in commonalities is generic and should live for more than one release. Here we're crafting an yaml for implementation so better to be straightforward. @jlurien @alpaycetin74 different view?

Yes, I find maximum=1 reasonable. Thank you.

**Comment by jlurien** [comment]
@bigludo7, as you are also adopting the subscription model to other APIs, what is the agreement in Commonalities? Are we supposed to use the common artifact in Commonalities, as it is, even with the options which are not applicable to the current version, or APIs can make adjustments? I suggested to remove protocols and credentialTypes which are not supported but PErdor told me that the agreement was to include all options from the beginning and specify errors for options not applicable.

**Comment by PedroDiez** [comment]
Hi all,

This is my understading about the agreement in commonalities: @bigludo7, @shilpa-padgaonkar, @rartych 

- [Artifact](https://github.com/camaraproject/Commonalities/blob/main/artifacts/camara-cloudevents/event-subscription-template.yaml), is a common baseline to be reused as-is in any API (in APIs with Instance-based (implicit) subscription, only applies sink and sinkCredential Model). That was the reason to model all the variants of Cloudevents and to add specific exceptions to address current simplifications. Benefit is that we will not need to update each time the model and avoid breaking-changes in the future.
- If the template is not followed, then it is not a template and probably this will generate confusion and misalignments between Subproyects

cc @maxl2287

**Comment by bigludo7** [comment]
Hi all,

I'm globally aligned with https://github.com/PedroDiez statement... but I should confess one contradiction (with myself ...ouch) because I was also fine to add this `max` set to 1 for the attribute `types` to be perfectly explicit to have only one event type per subscription for this release. 

For the protocol & tokenType I prefer to keep this in this yaml as defined in the template.

**Comment by jlurien** [comment]
> Hi all,
> 
> I'm globally aligned with https://github.com/PedroDiez statement... but I should confess one contradiction (with myself ...ouch) because I was also fine to add this `max` set to 1 for the attribute `types` to be perfectly explicit to have only one event type per subscription for this release.
> 
> For the protocol & tokenType I prefer to keep this in this yaml as defined in the template.

Indeed I read first the issue about adding the maximum even if it was not in the artifact, and I thought that the same criteria would apply to the rest of non-applicable stuff for this version :-). I think we should have a consistent approach for this, as other APIs may choose to not add the maximum, and that would create inconsistency.

IMO, commonalities versions should reflect what is applicable to each version, and they will evolve as we support more options, but if the agreement is to model the operations as in the template, we should stick to whatever it is in the artifact for this version.

**Comment by maxl2287** [comment]
@PedroDiez @jlurien @bigludo7 
What about setting a minimum of Items to 1?
`minItems: 1`


Because otherwise it would also be possible to provide an empty array (what is I guess not the expectation, when creating subscriptions)

**Comment by jlurien** [comment]
If that was the decision, I'm OK with keeping all unused values for protocol and credentialType. There are specific examples for errors INVALID_PROTOCOL and INVALID_CREDENTIAL for those. There is no specific code for the case when more that one event, or zero items, are sent, so it makes even more sense to add the limits minItems, maxItems in the schema, and rely on the generic 400 INVALID_ARGUMENT when a request body is not compliant.

**Comment by bigludo7** [comment]
Agreed also with @maxl2287 proposal for the `min`. I will take care of this one in Commonalities !

**Comment by maxl2287** [comment]
@jlurien & @bigludo7 ready to be reviewed again :)

**Review by bigludo7 — DISMISSED** [review]
LGTM/
For simplicity probably good to manage device identifier update defined in [PR213 ](https://github.com/camaraproject/Commonalities/pull/213)in another issue/PR.

**Review by bigludo7 — APPROVED** [review]
LGTM/

**Review by jlurien — APPROVED** [review]
LGTM

**Comment by bigludo7** [comment]
@maxl2287 We can merge this one, correct?

**Comment by maxl2287** [comment]
@bigludo7 I am finished. So yes. :)

**Pull request merged**